### DISCUSSION TITLE
OSHI supportability on Windows 2019 containers

### DISCUSSION THREAD

**Discussion by RajeshPutta87** [oshi/oshi#2271] [Q&A]
Hi OSHI Team,

We are leveraging OSHI to collect CPU information on Windows 2019 containers. We have noticed that CPU usage was always 100% and after deeper analysis we have noticed listed exception and we even tried to start our process using ContainerAdministrator user on container but with no luck.

Please note that this is how we use OSHI for CPU calculation in our product. Because values are always 0. [CPU busy %=100-CPU idle % which comes out always as 100% ]

Code Snippet

```java
SystemInfo si = new SystemInfo();
HardwareAbstractionLayer hal = si.getHardware();
CentralProcessor cpuInfo = hal.getProcessor();
 int coreCount = cpuInfo.getPhysicalProcessorCount();
 int processorCount = cpuInfo.getLogicalProcessorCount();
 CentralProcessor.ProcessorIdentifier cpuId = cpuInfo.getProcessorIdentifier();
 long[] loadTicks = cpuInfo.getSystemCpuLoadTicks();
 log.debug("OSHI library output (SystemCpuLoadTicks): " + Arrays.toString(loadTicks));
 long idleTicks = loadTicks[CentralProcessor.TickType.IDLE.getIndex()];
 long stealTicks = loadTicks[CentralProcessor.TickType.STEAL.getIndex()];
 long systemTicks = loadTicks[CentralProcessor.TickType.SYSTEM.getIndex()]
         + loadTicks[CentralProcessor.TickType.IRQ.getIndex()]
         + loadTicks[CentralProcessor.TickType.SOFTIRQ.getIndex()];
 long userTicks = loadTicks[CentralProcessor.TickType.USER.getIndex()];
 long timeStamp = idleTicks + stealTicks + systemTicks + userTicks;


 Map<String, WindowsRawData.CpuPerf> cpusPerfMap = ImmutableMap.of("_Total", WindowsRawData.CpuPerf.builder()
         .timestamp(timeStamp)
         .idleTicks(idleTicks)
         .stealTicks(stealTicks)
         .systemTicks(systemTicks)
         .userTicks(userTicks)
         .build());

 log.debug("OSHI library output (cpusPerf): " + rawData.getCpusPerf());
```

All those timestamp, idleTicks, stealTicks, systemTicks, userTicks are reported as 0's

DEBUG WindowsRawCollector - OSHI library output (cpusPerf): {_Total=WindowsRawData.CpuPerf(timestamp=0, idleTicks=0, stealTicks=0, systemTicks=0, userTicks=0)}

We do not have any problems on Windows 2022 container. We would like to ask you 

1) Whether OSHI is supported on windows 2019 container
2) If it is supported what are the mandatory requirements for seamless functioning on Windows 2019 container.  This issue was replicated on 3-4 different user setup's. So this is very highly likely that whoever uses OSHI on Windows 2019 containers, they are supposed to face same problem.

Looking forward to your thoughts.

```
ac1152567711==> [extension-scheduler-pool-0] 17 Oct 2022 15:21:21,129 ERROR HkeyPerformanceDataUtil - Unable to locate English counter names in registry Perflib 009. Counters may need to be rebuilt:
com.sun.jna.platform.win32.Win32Exception: The system cannot find the file specified.
        at com.sun.jna.platform.win32.Advapi32Util.registryGetStringArray(Advapi32Util.java:948) ~[jna-platform-5.8.0.jar:5.8.0 (b0)]
        at com.sun.jna.platform.win32.Advapi32Util.registryGetStringArray(Advapi32Util.java:923) ~[jna-platform-5.8.0.jar:5.8.0 (b0)]
        at com.sun.jna.platform.win32.Advapi32Util.registryGetStringArray(Advapi32Util.java:897) ~[jna-platform-5.8.0.jar:5.8.0 (b0)]
        at oshi.driver.windows.registry.HkeyPerformanceDataUtil.mapCounterIndicesFromRegistry(HkeyPerformanceDataUtil.java:301) ~[oshi-core-5.7.4.jar:5.7.4]
        at oshi.driver.windows.registry.HkeyPerformanceDataUtil.<clinit>(HkeyPerformanceDataUtil.java:69) ~[oshi-core-5.7.4.jar:5.7.4]
        at oshi.driver.windows.registry.ProcessPerformanceData.buildProcessMapFromRegistry(ProcessPerformanceData.java:69) ~[oshi-core-5.7.4.jar:5.7.4]
        at oshi.software.os.windows.WindowsOperatingSystem.queryProcessMapFromRegistry(WindowsOperatingSystem.java:333) ~[oshi-core-5.7.4.jar:5.7.4]
        at oshi.util.Memoizer$1.get(Memoizer.java:87) [oshi-core-5.7.4.jar:5.7.4]
        at oshi.software.os.windows.WindowsOperatingSystem.processMapToList(WindowsOperatingSystem.java:303) [oshi-core-5.7.4.jar:5.7.4]
        at oshi.software.os.windows.WindowsOperatingSystem.queryAllProcesses(WindowsOperatingSystem.java:265) [oshi-core-5.7.4.jar:5.7.4]
        at oshi.software.common.AbstractOperatingSystem.getProcesses(AbstractOperatingSystem.java:107) [oshi-core-5.7.4.jar:5.7.4]
        at oshi.software.os.OperatingSystem.getProcesses(OperatingSystem.java:271) [oshi-core-5.7.4.jar:5.7.4]
```
Thanks
Rajesh

**Comment by mprins** [comment]
> Whether OSHI is supported on windows 2019 container

I think you are the first to try. As you can see in the error message the required performance counters are missing in your container an you are being urged to rebuild those.
Also many fixes have been made wrt. windows perf counters since release of version 5.7.4 from may 2021 that your are using. You should consider 5.7.4 as unsupported, so please upgrade to the current release (6.3.0) and report back with your findings

**Comment by dbwiddis** [comment]
> Whether OSHI is supported on windows 2019 container

Broadly speaking, OSHI supports "Windows" and we make a best effort to enable whatever features we can when Windows is being operated in a virtual environment such as a VM or Container.  The unfortunate situation here is that we only have access to whatever data the VM/Container/Hypervisor provides for the host operating system.

Specific to your question, the stack trace shows that OSHI could not find English language performance counters in the registry at the expected standard location `HKEY_LOCAL_MACHINE\Software\Microsoft\Windows NT\CurrentVersion\Perflib\009`.  There are many ways these values may be corrupted, and this problem is common on systems which have been localized to use a different language, where the English counters get "erased" somehow.   

As @mprins noted above, there have been various localization-based bugs in OSHI trying to capture these or use fallback options such as WMI which are always in English.  The TLDR here: use OSHI 6.2.1 or later (current is 6.3.0) which finally fixes all these counter-based localization bugs (and provides much more detailed log messages if they still fail) and see if the issue persists (or you get more informative logs).

If you still have the problem, then it's an issue with the container having visibility of the performance counters.  OSHI attempts the following sequence to find the counters:
 - Uses the registry's `HKEY_PERFORMANCE_DATA` field to directly access all the counters in one pull of memory.  The Win32 Docs advise that it's preferred to use the Perfmon API instead, but this way is faster. :-)
 - If the registry method doesn't work, uses the Perfmon API.
 - Both of the above methods require localization steps. If they fail, OSHI attempts to read the counters from a WMI table that reports the identical counters; this tends to be a bit slower but is always English and tends to work around corrupt counter issues.

It is possible that none of the above methods work and you will get the symptoms you report, if the counters simply aren't enabled on your container.  I don't use Windows containers so it is difficult for me to advise further, but web searches using the keywords "windows container" and "performance counters" should offer some advice.  

There are different container images available.  The "Nano Server" images do not have performance counters, so at least the "Server Core" level image is required.

See also #1976 which reported similar issues.   There are instructions on enabling performance counters [here](https://docs.appdynamics.com/display/PRO14S/Enable+Monitoring+for+Windows+Performance+Counters) and you can peruse [these files](https://github.com/oshi/oshi/tree/master/oshi-core/src/main/java/oshi/driver/windows/perfmon) to identify which counters OSHI uses.

See also #2026 which noted [a Windows bug](https://docs.microsoft.com/en-us/troubleshoot/windows-server/virtualization/api-call-performance-counters-delay-container) impacting performance counters on Windows containers.  That link includes a workaround.

**Comment by dbwiddis** [comment]
This [thread](https://github.com/microsoft/ApplicationInsights-dotnet-server/issues/676) may or may not be useful as well.

**Comment by RajeshPutta87** [comment]
Hi ,

We have upgraded OSHI to 6.3.1 and still no use. We are still seeing 100% CPU usage. Please let us know if in case you have any other thoughts to address this issue

6e0208046be3==> [extension-scheduler-pool-0] 11 Nov 2022 05:06:19,655 ERROR HkeyPerformanceDataUtil - Unable to locate English counter names in registry Perflib 009. Counters may need to be rebuilt: 
com.sun.jna.platform.win32.Win32Exception: The system cannot find the file specified.

The same problem doesn't occur with Windows 2022 containers. This happens only with windows 2019 containers.

**Comment by dbwiddis** [comment]
Hey @RajeshPutta87 the direct source of that error message is that OSHI can not find performance counters at this location: `HKEY_LOCAL_MACHINE\Software\Microsoft\Windows NT\CurrentVersion\Perflib\009`.

Use your registry editor to look there yourself.  If the counters are missing, no software that accesses the performance counters will behave properly.

[This MSDN article](https://learn.microsoft.com/en-us/troubleshoot/windows-server/performance/rebuild-performance-counter-library-values) describes the issue.
> When you use the System Monitor tool, some counters may be missing or don't contain counter data. The base set of performance counter libraries may become corrupted and may need to be rebuilt. 

There are several steps there if you have extended the counters; if you are only using default counters the steps are simple (the uppercase R is important);
```
cd \windows\system32
lodctr /R
```

Check the above registry key again for those counters.  If they're still not there after you have rebuilt them, I'd suggest consulting a Windows support forum to further troubleshoot.

Performing a search using your favorite search engine using that registry key will give you many other sites describing similar symptoms and suggesting various fixes.  But the bottom line is:  _your performance counters are disabled or corrupt_, and no software, either OSHI or any other software, can read them unless you fix the underlying problem on the container.

**Comment by RajeshPutta** [comment]
@dbwiddis We tried to import Perflib0009 counters as listed here 
<img width="1789" alt="Screenshot 2022-12-16 at 2 02 40 PM" src="https://user-images.githubusercontent.com/1021671/208057138-4b525c69-9add-461f-a8f5-670d1b40c10a.png">
 Even after that there is no luck. Please find log details

bfe0eacc3e5e==> [extension-scheduler-pool-5] 16 Dec 2022 08:24:01,037 DEBUG WindowsRawCollector - OSHI library output (SystemCpuLoadTicks): [0, 0, 0, 0, 0, 0, 0, 0]
bfe0eacc3e5e==> [extension-scheduler-pool-11] 16 Dec 2022 08:24:26,361 DEBUG WindowsRawCollector - OSHI library output (SystemCpuLoadTicks): [0, 0, 0, 0, 0, 0, 0, 0]
bfe0eacc3e5e==> [extension-scheduler-pool-7] 16 Dec 2022 08:24:30,989 DEBUG WindowsRawCollector - OSHI library output (SystemCpuLoadTicks): [0, 0, 0, 0, 0, 0, 0, 0]
bfe0eacc3e5e==> [extension-scheduler-pool-10] 16 Dec 2022 08:24:56,302 DEBUG WindowsRawCollector - OSHI library output (SystemCpuLoadTicks): [0, 0, 0, 0, 0, 0, 0, 0]
bfe0eacc3e5e==> [extension-scheduler-pool-11] 16 Dec 2022 08:25:02,247 DEBUG WindowsRawCollector - OSHI library output (SystemCpuLoadTicks): [0, 0, 0, 0, 0, 0, 0, 0]

bfe0eacc3e5e==> [extension-scheduler-pool-9] 16 Dec 2022 08:23:30,947 DEBUG WindowsRawCollector - OSHI library output (cpusPerf): {_Total=WindowsRawData.CpuPerf(timestamp=0, idleTicks=0, stealTicks=0, systemTicks=0, userTicks=0)}
bfe0eacc3e5e==> [extension-scheduler-pool-7] 16 Dec 2022 08:23:56,261 DEBUG WindowsRawCollector - OSHI library output (cpusPerf): {_Total=WindowsRawData.CpuPerf(timestamp=0, idleTicks=0, stealTicks=0, systemTicks=0, userTicks=0)}
bfe0eacc3e5e==> [extension-scheduler-pool-5] 16 Dec 2022 08:24:01,037 DEBUG WindowsRawCollector - OSHI library output (cpusPerf): {_Total=WindowsRawData.CpuPerf(timestamp=0, idleTicks=0, stealTicks=0, systemTicks=0, userTicks=0)}
bfe0eacc3e5e==> [extension-scheduler-pool-11] 16 Dec 2022 08:24:26,361 DEBUG WindowsRawCollector - OSHI library output (cpusPerf): {_Total=WindowsRawData.CpuPerf(timestamp=0, idleTicks=0, stealTicks=0, systemTicks=0, userTicks=0)}
bfe0eacc3e5e==> [extension-scheduler-pool-7] 16 Dec 2022 08:24:30,989 DEBUG WindowsRawCollector - OSHI library output (cpusPerf): {_Total=WindowsRawData.CpuPerf(timestamp=0, idleTicks=0, stealTicks=0, systemTicks=0, userTicks=0)}
bfe0eacc3e5e==> [extension-scheduler-pool-10] 16 Dec 2022 08:24:56,302 DEBUG WindowsRawCollector - OSHI library output (cpusPerf): {_Total=WindowsRawData.CpuPerf(timestamp=0, idleTicks=0, stealTicks=0, systemTicks=0, userTicks=0)}
bfe0eacc3e5e==> [extension-scheduler-pool-11] 16 Dec 2022 08:25:02,247 DEBUG WindowsRawCollector - OSHI library output (cpusPerf): {_Total=WindowsRawData.CpuPerf(timestamp=0, idleTicks=0, stealTicks=0, systemTicks=0, userTicks=0)}

**Comment by RajeshPutta** [comment]
@dbwiddis Could you please share your thoughts for above update.

Thanks
Rajesh

**Comment by dbwiddis** [comment]
@RajeshPutta I am not certain of any solution here, I only know what I have seen when searching for similar symptoms.

Please take a look at this page:  https://learn.microsoft.com/en-us/azure/azure-monitor/app/performance-counters?tabs=net-core-new and also https://learn.microsoft.com/en-us/azure/cloud-services/diagnostics-performance-counters

It suggests two things to check:
 - Permissions issues for the app pool service account
   - `net localgroup "Performance Monitor Users" /add "IIS APPPOOL\NameOfYourPool"`
 - Configuring additional counters beyond the default:
   - Get a list of what's available: `Get-Counter -ListSet *`
   - Edit `ApplicationInsights.config` to add ones that aren't in the list that you need, like so (this is a partial list for testing; if this works I can suggest a complete list to add for other OSHI counters):
```xml
<Add Type="Microsoft.ApplicationInsights.Extensibility.PerfCounterCollector.PerformanceCollectorModule, Microsoft.AI.PerfCounterCollector">
  <Counters>
    <Add PerformanceCounter="\Processor Information(??APP_WIN32_PROC??)\% DPC Time"/>
    <Add PerformanceCounter="\Processor Information(??APP_WIN32_PROC??)\% Interrupt Time"/>
    <Add PerformanceCounter="\Processor Information(??APP_WIN32_PROC??)\% Privileged Time"/>
    <Add PerformanceCounter="\Processor Information(??APP_WIN32_PROC??)\% Processor Time"/>
    <Add PerformanceCounter="\Processor Information(??APP_WIN32_PROC??)\% User Time"/>
  </Counters>
</Add>
```
